auto_tests/toxav_mt_test.cc

```diff
@@ -0,0 +1,498 @@
+#include <algorithm>
+#include <atomic>
+#include <chrono>
+#include <cstdio>
+#include <cstring>
+#include <iostream>
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+#include "../testing/misc_tools.h"
+#include "../toxav/toxav.h"
+#include "../toxcore/tox.h"
+#include "check_compat.h"
+
+// Make lines shorter
+using Clock = std::chrono::high_resolution_clock;
+using TimePoint = std::chrono::time_point<Clock>;
+
+namespace {
+// Maximum amount of time in iterations to wait for bootstrapping and friend
+// connections to succeed.
+constexpr uint32_t MAX_BOOTSTRAP_ITERATIONS = 1000;
+
+struct Tox_Deleter {
+  void operator()(Tox *tox) const { tox_kill(tox); }
+};
+
+using Tox_Ptr = std::unique_ptr<Tox, Tox_Deleter>;
+
+struct ToxAV_Deleter {
+  void operator()(ToxAV *tox) const { toxav_kill(tox); }
+};
+
+using ToxAV_Ptr = std::unique_ptr<ToxAV, ToxAV_Deleter>;
+
+static void t_accept_friend_request_cb(Tox *m, const uint8_t *public_key, const uint8_t *data,
+                                       size_t length, void *userdata) {
+  if (length == 7 && std::memcmp("gentoo", data, 7) == 0) {
+    ck_assert(tox_friend_add_norequest(m, public_key, nullptr) != (uint32_t)-1);
+  }
+}
+
+std::vector<Tox_Ptr> prepareNetwork(uint32_t count) {
+  Tox_Err_New error;
+
+  // Temporary bootstrap node
+  std::printf("Created 1 instance of Tox as boostrap node\n");
+  Tox_Ptr bootstrap = Tox_Ptr(tox_new_log(nullptr, &error, nullptr));
+  ck_assert(error == TOX_ERR_NEW_OK);
+
+  uint8_t bootstrap_key[TOX_PUBLIC_KEY_SIZE];
+  tox_self_get_dht_id(bootstrap.get(), bootstrap_key);
+  const uint16_t bootstrap_port = tox_self_get_udp_port(bootstrap.get(), nullptr);
+
+  std::cout << "Bootstrapping " << std::to_string(count) << " Tox nodes" << std::endl;
+
+  std::vector<Tox_Ptr> toxes(count);
+
+  for (auto &tox : toxes) {
+    tox = Tox_Ptr(tox_new_log(nullptr, &error, nullptr));
+    ck_assert(error == TOX_ERR_NEW_OK);
+
+    tox_bootstrap(tox.get(), "localhost", bootstrap_port, bootstrap_key, nullptr);
+    tox_callback_friend_request(tox.get(), t_accept_friend_request_cb);
+  }
+
+  // Create fully meshed friend network
+  for (size_t i = 0; i < toxes.size(); ++i) {
+    uint8_t address[TOX_ADDRESS_SIZE];
+    tox_self_get_address(toxes[i].get(), address);
+
+    for (size_t j = i + 1; j < toxes.size(); ++j) {
+      Tox_Err_Friend_Add error_add;
+      tox_friend_add(toxes[j].get(), address, (const uint8_t *)"gentoo", 7, &error_add);
+      ck_assert(error_add == TOX_ERR_FRIEND_ADD_OK);
+    }
+  }
+
+  // temporarily add bootstrap node to end of toxes, so we can iterate all
+  toxes.push_back(std::move(bootstrap));
+
+  uint32_t bootstrap_iteration = 0;
```

Comment:
I'd move this down to just before the for-loop, and move the initialisation into the for-init-stmt.

Reply:
Done